lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -184,34 +186,37 @@ static MemOp cloneMemWithNewType(ImplicitLocOpBuilder *b, MemOp op,
     SmallVector<Attribute> portAnno;
     for (auto attr : newMem.getPortAnnotation(portIdx)) {
       if (auto subAnno = attr.dyn_cast<SubAnnotationAttr>()) {
-        auto minIdx = oldPortType.getIndexForFieldID(subAnno.getMinFieldID());
-        auto maxIdx = oldPortType.getIndexForFieldID(subAnno.getMaxFieldID());
-        if (minIdx != maxIdx) {
-          op.emitError("Cannot handle annotation ranges spanning fields");
+        auto targetIndex = oldPortType.getIndexForFieldID(subAnno.getFieldID());
+
+        // Apply annotations to all elements if the target is the whole
+        // sub-field.
+        if (subAnno.getFieldID() == oldPortType.getFieldID(targetIndex)) {
+          portAnno.push_back(SubAnnotationAttr::get(
+              b->getContext(), portType.getFieldID(targetIndex),
+              subAnno.getAnnotations()));
           continue;
         }
-        if (minIdx >= 3) { // data or mask
-          // is the annotation in the filtered field?
-          // FieldID of the annotation in the data field.
-          auto minID =
-              oldPortType.rootChildFieldID(subAnno.getMinFieldID(), minIdx);
-          // FieldID of the annotation in filtered data type.
-          minID = op.getDataType().rootChildFieldID(minID.first, field.index);
-          if (minID.second) { // contained!
+
+        // Handle `data` and `mask` sub-fields.
+        if (targetIndex >= 3) {
```

Comment:
Well, one follow-up issue.  (This shouldn't get fixed in this PR.)  Does this work for rw ports which have a type like follows:

```
!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, wmode: uint<1>, rdata flip: bundle<a: uint<8>, b: uint<8>>, wdata: bundle<a: uint<8>, b: uint<8>>, wmask: bundle<a: uint<1>, b: uint<1>>
```

Reply:
:smiley: Nope. This reminds me a question I have had: Is the order of these sub-fields fixed? Are there any optional sub-fields (such as 'en')? It seems this is not explicitly defined in the FIRRTL spec. I think the most robust way is a set of string comparisons with `data`, `rdata`, `wdata`, and `mask`. But if the order and presence can be guaranteed, we should be able to avoid the comparisons.

Reply:
It should be in sections 5.11.1--5.11.3.  The order is entirely fixed and needs to be because ordering matters for connect semantics (but not for partial connects).  Nothing is optional.

Note: that looking at the spec now, the types of the memory ports are actually listed as different from what we have here.  This may be a bug in CIRCT. 😅

Reply:
Nice, thanks! For this PR, I'll first use the order in CIRCT.